pkg/controller/volume/pvcprotection/pvc_protection_controller.go

```diff
@@ -418,3 +437,64 @@ func (c *Controller) enqueuePVCs(pod *v1.Pod, deleted bool) {
 		}
 	}
 }
+
+// This function checks if the PV correspodning to the PVC is still attached to
+// some node.
+func (c *Controller) isVolumeInUseByNode(pvc *v1.PersistentVolumeClaim) (bool, error) {
+	volName := pvc.Spec.VolumeName
+	// isVolumeInUseByNode is called only when PVC is a delete candidate. In the
+	// scenario where a PVC is already marked for deletion and the volume name is
+	// not stamped on the PVC, skip the nodes attachment check.
+	if volName == "" {
+		klog.Infof("Failed to fetch PV name for the PVC %s/%s which is marked for deletion, skip node attached check for PV", pvc.Namespace, pvc.Name)
+		return false, nil
+	}
+
+	// Check node status in the informer cache, if error encountered or volume not found, do not return immediately, try the API server for consistency.
+	if nodeName, ok := pvc.Annotations[AnnSelectedNode]; ok {
+		klog.V(3).Infof("For claim %s/%s, checking for volume %s in informer cache for selected node %s", pvc.Namespace, pvc.Name, volName, nodeName)
+		selectedNode, err := c.nodeLister.Get(nodeName)
+		nodes := []*v1.Node{selectedNode}
+		if err != nil {
+			klog.V(3).Infof("unexpected error getting target node %q from node informer for claim %s/%s: %v", nodeName, pvc.Namespace, pvc.Name, err)
+		} else if c.isVolumeInUse(volName, nodes) {
+			return true, nil
+		}
+	} else {
+		klog.V(3).Infof("For claim %s/%s, checking for volume %s in node informer cache", pvc.Namespace, pvc.Name, volName)
+		nodes, err := c.nodeLister.List(labels.Everything())
+		if err != nil {
+			return false, fmt.Errorf("unexpected error listing nodes from node informer: %v", err)
+		}
+		if c.isVolumeInUse(volName, nodes) {
+			return true, nil
+		}
+	}
+
+	klog.V(3).Infof("For claim %s/%s, volume %s, list nodes from API server", pvc.Namespace, pvc.Name, volName)
+	// If we reach here, we could not determine if the volume is attached to any node.
+	// Fetch node list from API server to check node.Status.VolumesInUse.
+	nodes, err := c.client.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return false, fmt.Errorf("List of nodes from API server failed: %s", err.Error())
+	}
+	nodesPtr := []*v1.Node{}
+	for i := range nodes.Items {
+		nodesPtr = append(nodesPtr, &nodes.Items[i])
+	}
+
+	return c.isVolumeInUse(volName, nodesPtr), nil
+}
+
+// Helper function to determine if a given volume is present in the node's volumesInUse list.
+func (c *Controller) isVolumeInUse(volName string, nodes []*v1.Node) bool {
+	for _, node := range nodes {
+		for _, volInUse := range node.Status.VolumesInUse {
+			if strings.Contains(string(volInUse), volName) {
```

Comment:
need to double check whether strings.contains always work..
do you know the volume format of VolumeInUse? I remember it is a unique key which contains some volume name.

Reply:
example entry in volumesInUse:
```
"kubernetes.io/csi/pd.csi.storage.gke.io^projects/mattcary-saikatroyc-test/zones/us-central1-b/disks/pvc-8198854a-e42e-4cb2-8d30-a911c3e7261e"
```
the part pvc-8198854a-e42e-4cb2-8d30-a911c3e7261e matches the pvc.Spec.VolumeName

Reply:
To be safer, better to use filter the last part, and then use exact match.

Reply:
Can we assume uniform volume handle naming for all PVs? "/" separated? then we can split by '/' and compare against the last part?